pkg/k8s/client/object_tracker.go

```diff
@@ -0,0 +1,617 @@
+		return nil, err
+	}
+	log.Debug("Get", logfieldResourceVersion, rev)
+	return obj.o, nil
+}
+
+// List implements testing.ObjectTracker.
+func (s *statedbObjectTracker) List(gvr schema.GroupVersionResource, gvk schema.GroupVersionKind, ns string, opts ...metav1.ListOptions) (runtime.Object, error) {
+	// Heuristic for list kind: original kind + List suffix. Might
+	// not always be true but this tracker has a pretty limited
+	// understanding of the actual API model.
+	listGVK := gvk
+	listGVK.Kind = listGVK.Kind + "List"
+	// GVK does have the concept of "internal version". The scheme recognizes
+	// the runtime.APIVersionInternal, but not the empty string.
+	if listGVK.Version == "" {
+		listGVK.Version = runtime.APIVersionInternal
+	}
+
+	list, err := s.scheme.New(listGVK)
+	if err != nil {
+		return nil, err
+	}
+
+	if !meta.IsListType(list) {
+		return nil, fmt.Errorf("%q is not a list type", listGVK.Kind)
+	}
+
+	var fieldSelector fields.Selector
+	if len(opts) > 0 {
+		opt := opts[0]
+		if opt.FieldSelector != "" {
+			fieldSelector, err = fields.ParseSelector(opt.FieldSelector)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	matchingObjects := []runtime.Object{}
+	txn := s.db.ReadTxn()
+
+	for obj, rev := range s.tbl.All(txn) {
+		if obj.domain != s.domain || obj.gvr != gvr ||
+			(ns != "" && obj.Namespace != ns) {
+			continue
+		}
+
+		if fieldSelector != nil {
+			if !objectMatchesFieldSelector(obj.o, fieldSelector) {
+				s.log.Debug("List: Skipping object due to FieldSelector mismatch",
+					logfields.K8sNamespace, ns,
+					logfields.Name, obj.Name,
+					logfieldFieldSelector, fieldSelector)
+				continue
+			}
+		}
+
+		o := obj.o.DeepCopyObject()
+		m, _ := meta.Accessor(o)
+		m.SetResourceVersion(strconv.FormatUint(rev, 10))
+		matchingObjects = append(matchingObjects, o)
+	}
+	m, _ := meta.ListAccessor(list)
+	m.SetResourceVersion(strconv.FormatUint(s.tbl.Revision(txn), 10))
+
+	if err := meta.SetList(list, matchingObjects); err != nil {
+		return nil, err
+	}
+
+	s.log.Debug(
+		"List",
+		logfieldClientset, s.domain,
+		logfieldGVR, gvr,
+		logfields.K8sNamespace, ns,
+		logfields.Count, len(matchingObjects),
+	)
+	return list, nil
+}
+
+// Patch implements testing.ObjectTracker.
+func (s *statedbObjectTracker) Patch(gvr schema.GroupVersionResource, obj runtime.Object, ns string, opts ...metav1.PatchOptions) error {
+	return s.Update(gvr, obj, ns)
+}
+
+// Update implements testing.ObjectTracker.
+func (s *statedbObjectTracker) Update(gvr schema.GroupVersionResource, obj runtime.Object, ns string, opts ...metav1.UpdateOptions) error {
+	obj = obj.DeepCopyObject()
+	newMeta, err := meta.Accessor(obj)
+	if err != nil {
+		s.log.Debug("Update", logfields.Error, err)
+		return err
+	}
+	if len(newMeta.GetNamespace()) == 0 {
+		newMeta.SetNamespace(ns)
+	}
+	wtxn := s.db.WriteTxn(s.tbl)
+	version := s.tbl.Revision(wtxn) + 1
+	newMeta.SetResourceVersion(strconv.FormatUint(version, 10))
```

Comment:
Do we somehow need to be careful not to increment the version if the object hasn't changed - as in, does Update even get called with unchanged objects?

Reply:
Ah good question. No idea. Will look into it.

Reply:
I think we don't need to care about this. For sure the client-go fake didn't care.